src/wasm/wasm-s-parser.cpp

```diff
@@ -538,11 +538,109 @@ FunctionType* SExpressionWasmBuilder::parseTypeRef(Element& s) {
     throw ParseException("invalid type reference", s.line, s.col);
   }
   IString name = getFunctionTypeName(*s[1]);
-  FunctionType* ft = wasm.getFunctionTypeOrNull(name);
-  if (!ft) {
+  FunctionType* functionType = wasm.getFunctionTypeOrNull(name);
+  if (!functionType) {
     throw ParseException("bad function type for import", s[1]->line, s[1]->col);
   }
-  return ft;
+  return functionType;
+}
+
+// Prases typeuse, a reference to a type definition. It is in the form of either
+// (type index) or (type name), possibly augmented by inlined (param) and
+// (result) nodes. (type) node can be omitted as well, in which case we get an
+// existing type if there's one with the same structure or create one.
+// Outputs are returned by parameter references.
+// typeuse ::= (type index|name)+ |
+//             (type index|name)+ (param ..)* (result ..)* |
+//             (param ..)* (result ..)*
+// TODO Remove FunctionType* parameter and the related logic to create
+// FunctionType once we remove FunctionType class.
+size_t SExpressionWasmBuilder::parseTypeUse(Element& s,
+                                            size_t startPos,
+                                            FunctionType*& functionType,
+                                            std::vector<NameType>& namedParams,
+                                            Type& result) {
+  size_t i = startPos;
+  bool typeExists = false, paramOrResultExists = false;
+  if (i < s.size() && elementStartsWith(*s[i], TYPE)) {
+    typeExists = true;
+    functionType = parseTypeRef(*s[i++]);
+  }
+  size_t paramPos = i;
+
+  size_t localIndex = 0;
+  while (i < s.size() && elementStartsWith(*s[i], PARAM)) {
+    paramOrResultExists = true;
+    auto newParams = parseParamOrLocal(*s[i++], localIndex);
+    namedParams.insert(namedParams.end(), newParams.begin(), newParams.end());
+  }
+  result = none;
+  if (i < s.size() && elementStartsWith(*s[i], RESULT)) {
+    paramOrResultExists = true;
+    result = parseResult(*s[i++]);
+  }
+
+  // verify if (type) and (params)/(result) match, if both are specified
+  if (typeExists && paramOrResultExists) {
+    size_t line = s[paramPos]->line, col = s[paramPos]->col;
+    const char* msg = "type and param/result don't match";
+    if (functionType->result != result) {
+      throw ParseException(msg, line, col);
+    }
+    if (functionType->params.size() != namedParams.size()) {
+      throw ParseException(msg, line, col);
+    }
+    for (size_t i = 0, n = namedParams.size(); i < n; i++) {
+      if (functionType->params[i] != namedParams[i].type) {
+        throw ParseException(msg, line, col);
+      }
+    }
+  }
+
+  // If (type) does not exist, check if there's a matching type, and if there
```

Comment:
I don't think this matches the spec behavior -- if a type use doesn't match it should be inserted at the end of the module (after all other types have been parsed). See http://webassembly.github.io/spec/core/text/modules.html#abbreviations

Reply:
You're right.. my code didn't change the order of type parsing and creation, so it was preexisting. But it's still good to fix that. Will fix it soon. Thanks!

Reply:
Oh by the way... when we add it to the end of type section, does the order matter? The current code creates all function types first. But one of them has `call_indirect` or something, should we add that type before adding the type for the next function? For example,

```wast
(func $a (param i32)
 (call_indirect (param i32 i32) ...)
)
(func $b (param i32 i32 i32) ...)
```

In this case, we currently add (and previously added) types in `i` -> `iii` -> `ii` order. But should we do that in `i` -> `ii`-> `iii` order? Unlike the 'type section first' fix, this will need some significant changes to the structure...

Reply:
Yes, the spec defines this too (with tests, IIRC):

> Abbreviations are expanded in the order they appear, such that previously inserted type definitions are reused by consecutive expansions.

Reply:
OTL

Reply:
Talked with @kripken and we agreed we may not need to fix it now; it's preexisting and the main purpose of this CL was to extract common routines for parsing typeuse.